[Inductor] Optimal Epilogue fusion overlapping with Async Pipelined Autotuning

PaulZhang12 opened the pull request:
Differential Revision: D89255517

Previously, we overlap the template fusion process with async_compile finely in https://github.com/pytorch/pytorch/pull/170565. This PR introduces properly overlapping getting the minimum timings from `get_min_choice` in the async pipelined autotuning mode, where the benchmarking happens on a different process. Furthermore, this PR properly enables epilogue/prologue fusion with async pipelined autotuning, with the limitation of no benchmarking epilogues. Previously, epilogue/prologue fusions were not compatible with async pipelined autotuning.

This brings compilation time for a representative internal arch down from ~23s (no async pipelined autotuning w/ multi template path) -> 19s (async pipelined autotuning)


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

eellison reviewed (commented):
looks good, couple questions

PaulZhang12 committed:
[Inductor] Optimal Epilogue fusion overlapping with Async Pipelined Autotuning (#171011)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/171011

Test Plan: buck2 run mode/opt caffe2/test/inductor:max_autotune -- -r TestMaxAutotuneAsyncPipelined

Reviewed By: wdvr

Differential Revision: D89255517

PaulZhang12 commented:
@pytorchbot merge -i
